### PR TITLE
Centralize & cleanup DefaultConfiguration creation logic

### PR DESCRIPTION
**Pull request by JLLeitschuh**
### Context

This cleans up the creation of the `DefaultConfiguration` to use the Factory pattern. This centralizes the creation logic in a single location and is an application of the dependency inversion principle.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 1]
```diff
@@ -1162,10 +1162,16 @@ private DefaultConfiguration newConfiguration() {
...
-        DefaultConfiguration copiedConfiguration = instantiator.newInstance(DefaultConfiguration.class, domainObjectContext, newName, configurationsProvider, resolver, listenerManager,
-            metaDataProvider, childResolutionStrategy, fileCollectionFactory, buildOperationExecutor, instantiator, artifactNotationParser, capabilityNotationParser,
-            attributesFactory, rootComponentMetadataBuilder, documentationRegistry, userCodeApplicationContext, owner, projectStateRegistry, workerThreadRegistry,
-            domainObjectCollectionFactory, calculatedValueContainerFactory);
+//        DefaultConfiguration copiedConfiguration = instantiator.newInstance(DefaultConfiguration.class, domainObjectContext, newName, configurationsProvider, resolver, listenerManager,
+//            metaDataProvider, childResolutionStrategy, fileCollectionFactory, buildOperationExecutor, instantiator, artifactNotationParser, capabilityNotationParser,
+//            attributesFactory, rootComponentMetadataBuilder, documentationRegistry, userCodeApplicationContext, owner, projectStateRegistry, workerThreadRegistry,
+//            domainObjectCollectionFactory, calculatedValueContainerFactory);
```
Comment (JLLeitschuh) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:1168]: I should remove this

**Comment by JLLeitschuh** [comment]
@bot-gradle test QFL

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java [hunk 2]
```diff
@@ -300,78 +300,84 @@ BaseRepositoryFactory createBaseRepositoryFactory(
...
-                                                                    ProjectStateRegistry projectStateRegistry,
-                                                                    CalculatedValueContainerFactory calculatedValueContainerFactory,
-                                                                    DocumentationRegistry documentationRegistry,
-                                                                    CollectionCallbackActionDecorator callbackDecorator,
-                                                                    UserCodeApplicationContext userCodeApplicationContext,
-                                                                    WorkerThreadRegistry workerThreadRegistry,
-                                                                    DomainObjectCollectionFactory domainObjectCollectionFactory,
-                                                                    NotationParser<Object, ComponentSelector> moduleSelectorNotationParser,
-                                                                    ObjectFactory objectFactory) {
+        ConfigurationContainerInternal createConfigurationContainer(
```
Comment (JLLeitschuh) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java:332]: This has a bunch of parameters that are no longer used and should be removed. I'm doing that now.

**Comment by JLLeitschuh** [comment]
@bot-gradle test QFL

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 3]
```diff
@@ -1162,10 +1162,12 @@ private DefaultConfiguration newConfiguration() {
         String newName = getNameWithCopySuffix();
 
         Factory<ResolutionStrategyInternal> childResolutionStrategy = resolutionStrategy != null ? Factories.constant(resolutionStrategy.copy()) : resolutionStrategyFactory;
-        DefaultConfiguration copiedConfiguration = instantiator.newInstance(DefaultConfiguration.class, domainObjectContext, newName, configurationsProvider, resolver, listenerManager,
-            metaDataProvider, childResolutionStrategy, fileCollectionFactory, buildOperationExecutor, instantiator, artifactNotationParser, capabilityNotationParser,
-            attributesFactory, rootComponentMetadataBuilder, documentationRegistry, userCodeApplicationContext, owner, projectStateRegistry, workerThreadRegistry,
-            domainObjectCollectionFactory, calculatedValueContainerFactory);
+        DefaultConfiguration copiedConfiguration = defaultConfigurationFactory.create(
```
Comment (tresat) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:1165]: Is it safe to hold a reference to the factory like this and assume it can be used to create a copy with a new name?  Is the factory meant to be fully immutable?  Could any of the fields used as arguments when creating the new instance of `DefaultConfiguration` which now live in the factory class have changed between the construction of `this` and the call to `newConfiguration`, resulting a copy which isn't actually identical in the expected ways?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationFactory.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    /**
+     * Creates a new {@link DefaultConfiguration} instance.
+     */
+    DefaultConfiguration create(
+        String name,
+        ConfigurationsProvider configurationsProvider,
+        Factory<ResolutionStrategyInternal> resolutionStrategyFactory,
+        RootComponentMetadataBuilder rootComponentMetadataBuilder
+    ) {
+        ListenerBroadcast<DependencyResolutionListener> dependencyResolutionListeners =
```
Comment (tresat) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationFactory.java:114]: Why is this stored as a local instead of created inline in the `newInstance` call?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java [hunk 5]
```diff
@@ -186,10 +134,7 @@ private void copyAllTo(DefaultConfiguration detachedConfiguration, Dependency[]
     }
 
     private DefaultConfiguration newConfiguration(String name, ConfigurationsProvider detachedConfigurationsProvider, RootComponentMetadataBuilder componentMetadataBuilder) {
-        return instantiator.newInstance(DefaultConfiguration.class, context, name, detachedConfigurationsProvider, resolver, listenerManager,
-            dependencyMetaDataProvider, resolutionStrategyFactory, fileCollectionFactory, buildOperationExecutor, instantiator, artifactNotationParser,
-            capabilityNotationParser, attributesFactory, componentMetadataBuilder, documentationRegistry, userCodeApplicationContext,
-            context, projectStateRegistry, workerThreadRegistry, domainObjectCollectionFactory, calculatedValueContainerFactory);
+        return defaultConfigurationFactory.create(name, detachedConfigurationsProvider, resolutionStrategyFactory, componentMetadataBuilder);
```
Comment (tresat) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java:137]: Similar comment here as in `DefaultConfiguration` it is safe to depend on the values of the fields which now live in the factory holding their expected values?

Previously, a client of this class could keep a reference to any of these fields when constructing a new `DefaultConfigurationContainer` and modify them externally (if they are mutable, I didn't look very deeply) between the construction of `this` and the call to `newConfiguration`, as this class doesn't make defensive copies or anything.  This might break the expectations for how `newConfiguration` is meant to work.  If that was a problem, it would previously have been contained entirely within this class - a violation of an implicit class invariant - and now it would be distributed between this class and the factory.  Not even sure if this is a real problem, just something to consider.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java [hunk 6]
```diff
@@ -202,6 +200,7 @@ public DependencyResolutionScopeServices(DomainObjectContext domainObjectContext
 
         void configure(ServiceRegistration registration) {
             registration.add(DefaultTransformedVariantFactory.class);
+            registration.add(DefaultConfigurationFactory.class);
```
Comment (tresat) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java:203]: Any need to mark the new factory class @ThreadSafe like the other factory here?

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationSpec.groovy [hunk 7]
```diff
@@ -1763,11 +1765,30 @@ All Artifacts:
         _ * domainObjectContext.buildPath >> Path.path(buildPath)
         _ * domainObjectContext.model >> RootScriptDomainObjectContext.INSTANCE
 
-        def publishArtifactNotationParser = NotationParserBuilder.toType(ConfigurablePublishArtifact).toComposite()
-        new DefaultConfiguration(domainObjectContext, confName, configurationsProvider, resolver, listenerManager, metaDataProvider,
-            Factories.constant(resolutionStrategy), TestFiles.fileCollectionFactory(), new TestBuildOperationExecutor(), instantiator,
-            publishArtifactNotationParser, Stub(NotationParser), immutableAttributesFactory, rootComponentMetadataBuilder, Stub(DocumentationRegistry),
-            userCodeApplicationContext, domainObjectContext, projectStateRegistry, Stub(WorkerThreadRegistry), TestUtil.domainObjectCollectionFactory(), calculatedValueContainerFactory)
+        def publishArtifactNotationParser = new PublishArtifactNotationParserFactory(
```
Comment (tresat) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationSpec.groovy:1768]: Why the change in how `publishArtifactNotationParser` is created?

**Review by tresat — COMMENTED** [review]
Looks like a proper application of the factory pattern to me.  I'm still a newcomer to this area though.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java (hunk 5 above)
Comment (JLLeitschuh) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java:137]: If such expectations were expected to occur, I would presume that a test would ensure that behaviour continued to be maintained.

That being said, I did my best to be careful to ensure that the types that were fixed were passed as constructor arguments to the `DefaultConfigurationFactory`, and the things that varied (were manipulated) were passed to `create`.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationFactory.java (hunk 4 above)
Comment (JLLeitschuh) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationFactory.java:114]: Readability and potentially better stack traces. Personally, I don't care, but this seemed to read better IMHO.

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationSpec.groovy (hunk 7 above)
Comment (JLLeitschuh) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationSpec.groovy:1768]: Because I need an instance of `PublishArtifactNotationParserFactory` instead of `Factory<NotationParser<Object, ConfigurablePublishArtifact>>`. Unfortunately, the Gradle DI engine can't locate an instance of `Factory<NotationParser<Object, ConfigurablePublishArtifact>>` so I have to use the exact type `PublishArtifactNotationParserFactory`. I spent some time looking at how our service loader logic worked and toyed with the idea of trying to fix it to support locating and injecting a type matching `Factory<NotationParser<Object, ConfigurablePublishArtifact>>` but it seemed like a rabbit hole I didn't want to go down. As such, it was easier to just pass the concrete type here in the tests.

I agree that your proposal would be better.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java (hunk 6 above)
Comment (JLLeitschuh) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java:203]: Sure, why not. The class should be thread safe, it's just a factory.

Comment (JLLeitschuh) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java:203]: Done!

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 8]
```diff
@@ -241,28 +236,33 @@
...
-                                FileCollectionFactory fileCollectionFactory,
-                                BuildOperationExecutor buildOperationExecutor,
-                                Instantiator instantiator,
-                                NotationParser<Object, ConfigurablePublishArtifact> artifactNotationParser,
-                                NotationParser<Object, Capability> capabilityNotationParser,
-                                ImmutableAttributesFactory attributesFactory,
-                                RootComponentMetadataBuilder rootComponentMetadataBuilder,
-                                DocumentationRegistry documentationRegistry,
-                                UserCodeApplicationContext userCodeApplicationContext,
-                                DomainObjectContext owner,
```
Comment (JLLeitschuh) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:261]: Removed this argument as it was redundant.

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainerSpec.groovy [hunk 9]
```diff
@@ -77,10 +76,41 @@ class DefaultConfigurationContainerSpec extends Specification {
...
+        listenerManager,
+        metaDataProvider,
+        domainObjectContext,
+        fileCollectionFactory,
+        buildOperationExecutor,
+        Stub(PublishArtifactNotationParserFactory),
+        immutableAttributesFactory,
+        documentationRegistry,
+        userCodeApplicationContext
+        ,
```
Comment (JLLeitschuh) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainerSpec.groovy:91]: Wacky, this shouldn't be this way.

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainerTest.groovy [hunk 10]
```diff
@@ -70,32 +72,45 @@ class DefaultConfigurationContainerTest extends Specification {
...
+        buildOperationExecutor,
+        new PublishArtifactNotationParserFactory(
+            instantiator,
+            metaDataProvider,
+            taskResolver
+        ),
+        immutableAttributesFactory,
+        documentationRegistry,
+        userCodeApplicationContext
+        ,
```
Comment (JLLeitschuh) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainerTest.groovy:92]: IJ's "safe delete" does not do what I'd expect it to do 🤦

**Comment by JLLeitschuh** [comment]
@bot-gradle test QFL

**Comment by big-guy** [comment]
@bot-gradle help

**Comment by big-guy** [comment]
@bot-gradle test AST

**Comment by JLLeitschuh** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge
